### PR TITLE
Cleanup of slot-update-related content

### PR DESCRIPTION
**Pull request by PetarMax**
~~This PR moves the slot-update auxiliary lemmas out of the symbolic slot updates module so that the LLVM backend can use them. It also removes several no-longer-required related lemmas and claims.~~

EDIT: Adding more clarity below.

This PR is an off-shoot of #2604. The idea behind that PR was to consolidate all of the `#asWord`-related lemmas, also removing unneeded ones. As that was happening, an issue arose with the slot update lemmas, in that the related auxiliary functions were imported under a `symbolic` module, meaning that they would never get to the LLVM backend. However, these auxiliary functions are only meant to be used with concrete arguments, and when the booster sees a function with all arguments concrete, it passes it for evaluation to the LLVM backend - this is one of the key speed-ups. When the LLVM backend receives a function that it doesn't have, it throws an error and stops the simplificaiton process. So, the booster was effectively unable to apply almost any of the slot update lemmas, which were really applied only by Kore, supported by the lemmas that are subsumed by the slot update lemmas, but were not yet removed.

With all of the changes made, #2604 became too cumbersome to review integrally. This PR only:
- moves the auxiliary functions to a non-symbolic module (`word.md`), which seems appropriate as these primarily deal with words;
- removes the lemmas that we know are subsumed by the slot update lemmas; and
- temporarily comments out claims for which the new `#asWord` lemmas are needed - all of these claims pass in #2604.

### CODE REVIEW

File: kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/bitwise-simplification.k [hunk 1]
```diff
@@ -60,39 +60,6 @@ module BITWISE-SIMPLIFICATION [symbolic]
        andBool 0 <=Int Y andBool Y <Int pow256 -Int X
       [simplification, concrete(X), preserves-definedness]
 
-    // Generalization of: maxUIntXXX &Int #asWord ( BA )
```
Comment (PetarMax) [inline kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/bitwise-simplification.k:63]: These are subsumed by the slot update lemmas.

File: tests/specs/benchmarks/functional-spec.k [hunk 2]
```diff
@@ -85,12 +85,12 @@ module FUNCTIONAL-SPEC
        andBool #rangeBytes(32, SIGS)
        andBool notBool #ecrecEmpty(#bufStrict(32, HASH), #bufStrict(32, SIGV), #bufStrict(32, SIGR), #bufStrict(32, SIGS))
 
-    claim <k> runLemma(maxUInt160 &Int #asWord(#ecrec(#bufStrict(32, HASH), #bufStrict(32, SIGV), #bufStrict(32, SIGR), #bufStrict(32, SIGS)))) => doneLemma(#asWord(#ecrec(#bufStrict(32, HASH), #bufStrict(32, SIGV), #bufStrict(32, SIGR), #bufStrict(32, SIGS)))) ... </k>
-      requires #rangeUInt(256, HASH)
-       andBool #rangeUInt(8, SIGV)
-       andBool #rangeBytes(32, SIGR)
-       andBool #rangeBytes(32, SIGS)
-       andBool notBool #ecrecEmpty(#bufStrict(32, HASH), #bufStrict(32, SIGV), #bufStrict(32, SIGR), #bufStrict(32, SIGS))
+    // claim <k> runLemma(maxUInt160 &Int #asWord(#ecrec(#bufStrict(32, HASH), #bufStrict(32, SIGV), #bufStrict(32, SIGR), #bufStrict(32, SIGS)))) => doneLemma(#asWord(#ecrec(#bufStrict(32, HASH), #bufStrict(32, SIGV), #bufStrict(32, SIGR), #bufStrict(32, SIGS)))) ... </k>
```
Comment (PetarMax) [inline tests/specs/benchmarks/functional-spec.k:88]: Requires new `#asWord` lemmas from the upcoming PR.

File: tests/specs/mcd/flopper-file-pass-rough-spec.k [hunk 3]
```diff
@@ -3,127 +3,127 @@ requires "verification.k"
...
-       andBool ACCT_ID =/=Int 0
-       andBool #notPrecompileAddress(ACCT_ID)
-       andBool #rangeAddress(CALLER_ID)
-       andBool #rangeAddress(ORIGIN_ID)
-       andBool #rangeUInt(256, TIME)
-       andBool #rangeUInt(256, ACCT_ID_balance)
-       andBool VCallDepth <=Int 1024
-       andBool #rangeUInt(256, VCallValue)
-       andBool #rangeUInt(256, VChainId)
+    // // Flopper_file
```
Comment (PetarMax) [inline tests/specs/mcd/flopper-file-pass-rough-spec.k:6]: Requires new `#asWord` lemmas from the upcoming PR.

File: tests/specs/mcd-structured/verification.k [hunk 4]
```diff
@@ -265,7 +265,7 @@ module LEMMAS-MCD [symbolic]
   // Arithmetic
   // ########################
 
-    rule X <=Int maxUInt48 &Int (X +Int Y) => false
+    rule X <=Int #asWord ( #range ( #buf ( 32, X +Int Y ), 26, 6 ) ) => false
```
Comment (PetarMax) [inline tests/specs/mcd-structured/verification.k:268]: The `maxUInt48 &Int (X +Int Y)` term gets transformed into the other one, so this very specific simplification needs to follow.

File: tests/specs/mcd-structured/verification.k [hunk 5]
```diff
@@ -365,7 +365,7 @@ module LEMMAS-MCD [symbolic]
     rule chop(X *Int Y) /sWord ( Y +Int pow256 ) => X requires #rangeUInt(256, X) andBool           Y <Int 0 andBool #rangeSInt(256, X *Int Y)                    [simplification]
     rule chop(X *Int Y) /sWord chop(Y)           => X requires #rangeUInt(256, X) andBool #rangeSInt(256, Y) andBool #rangeSInt(256, X *Int Y) andBool Y =/=Int 0 [simplification]
 
-    rule maxUInt48 &Int (X +Int Y) <Int X => true
+    rule #asWord ( #range ( #buf ( 32, X +Int Y ), 26, 6 ) ) <Int X => true
```
Comment (PetarMax) [inline tests/specs/mcd-structured/verification.k:368]: Same as above.

File: tests/specs/functional/lemmas-spec.k [hunk 6]
```diff
@@ -504,7 +504,6 @@ module LEMMAS-SPEC
  // bitwise
  // -------
 
-    claim [bit-n005-0]: <k> runLemma ( notMaxUInt5   &Int X ) => doneLemma( 0 ) ... </k> requires #rangeUInt (   5 , X )
```
Comment (PetarMax) [inline tests/specs/functional/lemmas-spec.k:507]: The four `005` lemmas are not needed, as the illustrative examples are about manipulation in multiples of 8.

File: tests/specs/mcd/flopper-file-pass-rough-spec.k [hunk 7]
```diff
@@ -3,127 +3,127 @@ requires "verification.k"
 module FLOPPER-FILE-PASS-ROUGH-SPEC
     imports VERIFICATION
 
-    // Flopper_file
```
Comment (PetarMax) [inline tests/specs/mcd/flopper-file-pass-rough-spec.k:6]: Same as for the structured version above.

File: .github/workflows/test-pr.yml [hunk 8]
```diff
@@ -96,8 +96,8 @@ jobs:
           - name: 'Functional'
             test-suite: 'test-prove-functional'
             test-args:
-            timeout: 45
-            parallel: 2
+            timeout: 60
+            parallel: 1
```
Comment (ehildenb) [inline .github/workflows/test-pr.yml:100]: I think these changes are being subsumed here: https://github.com/runtimeverification/evm-semantics/pull/2608

File: kevm-pyk/src/tests/integration/test_prove.py [hunk 9]
```diff
@@ -329,7 +329,6 @@ def test_prove_rules(
     tmp_path: Path,
     caplog: LogCaptureFixture,
     no_use_booster: bool,
-    force_sequential: bool,
```
Comment (ehildenb) [inline kevm-pyk/src/tests/integration/test_prove.py:332]: I guess these changes are subsumed here: https://github.com/runtimeverification/evm-semantics/pull/2608

And can be removed from this PR.

File: tests/specs/functional/lemmas-spec.k [hunk 10]
```diff
@@ -532,33 +530,31 @@ module LEMMAS-SPEC
...
+    // claim [bit-160-nm-or-m]: <k> runLemma ( notMaxUInt160 &Int ( X |Int ( maxUInt160 &Int T ) ) ) => doneLemma( 0 ) ... </k> requires #rangeUInt (  96 , X ) andBool #rangeUInt(256, T)
+    // claim [bit-192-nm-or-m]: <k> runLemma ( notMaxUInt192 &Int ( X |Int ( maxUInt192 &Int T ) ) ) => doneLemma( 0 ) ... </k> requires #rangeUInt (  64 , X ) andBool #rangeUInt(256, T)
+    // claim [bit-208-nm-or-m]: <k> runLemma ( notMaxUInt208 &Int ( X |Int ( maxUInt208 &Int T ) ) ) => doneLemma( 0 ) ... </k> requires #rangeUInt (  48 , X ) andBool #rangeUInt(256, T)
+    // claim [bit-224-nm-or-m]: <k> runLemma ( notMaxUInt224 &Int ( X |Int ( maxUInt224 &Int T ) ) ) => doneLemma( 0 ) ... </k> requires #rangeUInt (  32 , X ) andBool #rangeUInt(256, T)
+    // claim [bit-240-nm-or-m]: <k> runLemma ( notMaxUInt240 &Int ( X |Int ( maxUInt240 &Int T ) ) ) => doneLemma( 0 ) ... </k> requires #rangeUInt (  16 , X ) andBool #rangeUInt(256, T)
+    // claim [bit-248-nm-or-m]: <k> runLemma ( notMaxUInt248 &Int ( X |Int ( maxUInt248 &Int T ) ) ) => doneLemma( 0 ) ... </k> requires #rangeUInt (   8 , X ) andBool #rangeUInt(256, T)
```
Comment (ehildenb) [inline tests/specs/functional/lemmas-spec.k:557]: These are commented intentionally?

File: kevm-pyk/src/kevm_pyk/kproj/evm-semantics/word.md [hunk 11]
```diff
@@ -585,6 +585,81 @@ Range of types
...
+The following auxiliary functions support the associated automated reasoning.
+
+-   `#getFirstOneBit` returns the 0-based index of the first non-zero bit of a word
+
+```k
+    syntax Int ::=  #getFirstOneBit(Int) [function, total]
+
+    rule [gfo-succ]: #getFirstOneBit(X:Int) => log2Int ( X &Int ( ( maxUInt256 xorInt X ) +Int 1 ) ) requires #rangeUInt(256, X) andBool X =/=Int 0 [preserves-definedness]
+    rule [gfo-fail]: #getFirstOneBit(_:Int) => -1                                                    [owise]
+```
```
Comment (ehildenb) [inline kevm-pyk/src/kevm_pyk/kproj/evm-semantics/word.md:603]: I'm not sure why these definitions are moved to this file. Do we need them in the LLVM backend sometimes? Maybe when the booster is sending terms to the LLVM backend?

File: kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/bitwise-simplification.k (hunk 1 above)
Comment (ehildenb) [reply kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/bitwise-simplification.k:63]: It seems that deleting these lemmas is the bulk of this change. Does just deleting these lemmas, and commenting the handful of proofs, suffice for this PR? Or do we also need to move the slot update reasoning module stuff?

File: .github/workflows/test-pr.yml (hunk 8 above)
Comment (PetarMax) [reply .github/workflows/test-pr.yml:100]: They will go away, yes.

File: tests/specs/functional/lemmas-spec.k (hunk 10 above)
Comment (PetarMax) [reply tests/specs/functional/lemmas-spec.k:557]: Yes.

File: kevm-pyk/src/kevm_pyk/kproj/evm-semantics/word.md (hunk 11 above)
Comment (PetarMax) [reply kevm-pyk/src/kevm_pyk/kproj/evm-semantics/word.md:603]: Will explain in PR info shortly.

File: kevm-pyk/src/tests/integration/test_prove.py (hunk 9 above)
Comment (PetarMax) [reply kevm-pyk/src/tests/integration/test_prove.py:332]: They will go away, yes.

File: kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/bitwise-simplification.k (hunk 1 above)
Comment (PetarMax) [reply kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/bitwise-simplification.k:63]: It's a bit the other way round, will explain in PR info.

**Review by ehildenb — APPROVED** [review]

**Pull request merged**